icicle/tests/test_hash_api.cpp

```diff
@@ -313,38 +316,210 @@ class HashSumBackend : public HashBackend
+  auto config = default_merkle_tree_config();
+  auto layer0_hash = HashSumBackend::create(5 * leaf_size, 2 * leaf_size); // in 5 leaves, out 2 leaves  200B ->  80B
+  auto layer1_hash = HashSumBackend::create(4 * leaf_size, leaf_size);     // in 4 leaves, out 1 leaf    80B  ->  20B
+  auto layer2_hash = HashSumBackend::create(leaf_size, leaf_size);         // in 1 leaf, out 1 leaf      20B  ->  20B
+  auto layer3_hash = HashSumBackend::create(5 * leaf_size, leaf_size); // in 5 leaves, out 1 leaf    20B  ->  4B output
+
+  std::vector<Hash> hashes = {layer0_hash, layer1_hash, layer2_hash, layer3_hash};
+
+  int output_store_min_layer;
+  randomize(&output_store_min_layer, 1);
+  output_store_min_layer = output_store_min_layer & 3; // Ensure index is in a valid 0-3 range
+  ICICLE_LOG_DEBUG << "Min store layer:\t" << output_store_min_layer;
+
+  auto prover_tree = MerkleTree::create(hashes, sizeof(merkel_hash_t), output_store_min_layer);
+
+  auto verifier_tree = MerkleTree::create(hashes, sizeof(merkel_hash_t), output_store_min_layer);
+
+  // build tree
+
+  // START_TIMER(MerkleTree_build)
+  ICICLE_CHECK(prover_tree.build(leaves, nof_leaves, config));
+  assert_valid_tree<uint32_t>(prover_tree, nof_leaves, leaves, hashes, config);
+  // std::cout << "MerkleTree build time: ";
+  // END_TIMER(MerkleTree_build, std::cout, true)
+
+  // get root and merkle-path to an element
+  int nof_idx_to_check = 5;
+  uint64_t leaf_indices[nof_idx_to_check];
+  randomize(leaf_indices, nof_idx_to_check);
+
+  for (int i = 0; i < nof_idx_to_check; i++) {
+    int leaf_idx = leaf_indices[i] % nof_leaves;
+
+    auto [root, root_size] = prover_tree.get_merkle_root();
+    MerkleProof merkle_proof{};
+    ICICLE_CHECK(prover_tree.get_merkle_proof(leaves, nof_leaves, leaf_idx, false, config, merkle_proof));
+
+    // Test valid proof
+    bool verification_valid = false;
+    ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+    ASSERT_TRUE(verification_valid);
+
+    // Test invalid proof (By modifying random data in the leaves)
+    verification_valid = true;
+    ICICLE_CHECK(prover_tree.get_merkle_proof(other_leaves, nof_leaves, leaf_idx, false, config, merkle_proof));
+    ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+    ASSERT_FALSE(verification_valid);
+
+    // Same for pruned proof
+    verification_valid = false;
+    ICICLE_CHECK(prover_tree.get_merkle_proof(leaves, nof_leaves, leaf_idx, true, config, merkle_proof));
+    ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+    ASSERT_TRUE(verification_valid);
+
+    // Test invalid proof (By adding random data to the proof)
+    verification_valid = true;
+    ICICLE_CHECK(prover_tree.get_merkle_proof(other_leaves, nof_leaves, leaf_idx, true, config, merkle_proof));
+    ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+    ASSERT_FALSE(verification_valid);
   }
+}
+
+TEST_F(HashApiTest, MerkleTree)
```

Comment:
what is the difference between the two tests? it's almost the same.
Can you make a test that scales with input size? So that we can easily test execution time on CPU and GPU.